### PR TITLE
fix(get): propagate next page size param #9029

### LINKED ISSUE #9029: Admin API query parameter `size` is not reported in the `next` URL in paginated lists
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

2.1.8

### Current Behavior

When doing the following request:

    GET http://localhost:8001/services?tags=test&size=1

The response contains a single item as expected by the `size` argument (I could not find it documented in the Kong Admin API, anyway)

but in the `next` field the contained URL is:

    "next": "/services?offset=WyIzZWQ2N2EwMC1mNGY2LTQ4NTUtYmQ0Yi0zMmZmMWMzNWVmNmMiXQ&tags=test" 

the `tags` parameter is present, but `size` is missing.

### Expected Behavior

The `size` parameter should have been present so that the pagination will keep respecting the size the user requested:

    "next": "/services?offset=WyIzZWQ2N2EwMC1mNGY2LTQ4NTUtYmQ0Yi0zMmZmMWMzNWVmNmMiXQ&tags=test" 

### Steps To Reproduce

```markdown
1. Do the request: `GET http://localhost:8001/services?tags=test&size=1`
2. Verify that the `next` field in the response lacks the `size` parameter
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jacksjm**
### Summary

Fixed to propagate size parameter to perform next request

### Full changelog

* Fixed to propagate the size parameter to perform the next request, ensuring the same size in requests

### Issue reference

Fix #9029

### CODE REVIEW

File: kong/api/endpoints.lua [hunk 1]
```diff
@@ -303,12 +303,17 @@ end
...
     local next_page_tags = ""
+    local next_page_size = ""
 
     local args = self.args.uri
     if args.tags then
       next_page_tags = "&tags=" .. escape_uri(type(args.tags) == "table" and args.tags[1] or args.tags)
     end
 
+    if args.size then
+      next_page_size = "&size=" .. escape_uri(type(args.size) == "number" and args.size[1] or args.size)
```
Comment (chronolaw) [inline kong/api/endpoints.lua:314]: It is a logic error, we can not access `[1]` of `number` type.

**Review by chronolaw — CHANGES_REQUESTED** [review]

File: kong/api/endpoints.lua (hunk 1 above)
Comment (jacksjm) [reply kong/api/endpoints.lua]: @chronolaw Sorry, my mistake! I validated now only if the type is numeric, I think it should solve

**Comment by chronolaw** [comment]
Great Job, Thank you Jackson. 

Because I am not familiar with admin api, I think that we need others to take a look then approve.

**Comment by jacksjm** [comment]
> Great Job, Thank you Jackson.
> 
> Because I am not familiar with admin api, I think that we need others to take a look then approve.

this is OK. could you tag a sponsor to review? I do not know them. Do you need to remove your change request?

**Review by chronolaw — APPROVED** [review]

File: kong/api/endpoints.lua [hunk 2]
```diff
@@ -303,12 +303,17 @@ end
...
   return not foreign_schema and function(self, db, helpers)
     local next_page_tags = ""
+    local next_page_size = ""
 
     local args = self.args.uri
     if args.tags then
       next_page_tags = "&tags=" .. escape_uri(type(args.tags) == "table" and args.tags[1] or args.tags)
     end
 
+    if (type(args.size) == "number") then
```
Comment (chronolaw) [inline kong/api/endpoints.lua:313]: style: this `if` condition should not use `(...)`.

Comment (jacksjm) [reply kong/api/endpoints.lua]: wow, ok, removed :smile:

**Comment by jacksjm** [comment]
@mayocream would you be on the admin api team for review?

**Review by mayocream — CHANGES_REQUESTED** [review]
Hi, although it’s a small change, but a proper test case should be written to ensure the functionality works.

**Comment by jacksjm** [comment]
> Hi, although it’s a small change, but a proper test case should be written to ensure the functionality works.

Thanks for your feedback @mayocream!!! I searched the admin_api tests for next_page validators but couldn't find them. I believe there is no validation on this return, if I am wrong please correct me. Do you have any validation suggestions for this scenario?

**Review by mayocream — DISMISSED** [review]
Should be in this format:
```diff
< local next_page = offset and fmt("/%s?offset=%s%s",
---
> local next_page = offset and fmt("/%s?offset=%s%s%s",
```

**Comment by mayocream** [comment]
I think this would be a feature (although very small); I'll include @sumimakito in this topic to see if it is good for Kong Manager (UI).

**Comment by sumimakito** [comment]
@jacksjm Thank you for your work 🙌
@mayocream Yes - I believe this would benefit KM and other projects that rely on the `next` link in the response, as long as we can ensure this change would not break them.

**Comment by jacksjm** [comment]
@mayocream fixed this :smile:

File: kong/api/endpoints.lua [hunk 3]
```diff
@@ -304,22 +304,28 @@ end
...
     end
 
-    local next_page = offset and fmt("/%s?offset=%s%s",
+    local next_page = offset and fmt("/%s?offset=%s%s%s",
                                      schema.admin_api_name or
                                      schema.name,
                                      escape_uri(offset),
-                                     next_page_tags) or null
+                                     next_page_tags,
+                                     next_page_size) or null
```
Comment (hbagdi) [inline kong/api/endpoints.lua:328]: Can you please include a test for this?

Comment (jacksjm) [reply kong/api/endpoints.lua:328]: Of course! But as I commented earlier, I didn't find any other automated tests for the next_page property. As the tests are segregated, I'm not sure how to proceed and in which file to put this test. If you direct me, I do it quickly :smile:

Comment (hbagdi) [reply kong/api/endpoints.lua:328]: Please take a look at the tests in https://github.com/Kong/kong/tree/master/spec/02-integration/04-admin_api and add in a test that asserts that the response has the appropriate field set to the value that you expect.

Comment (jacksjm) [reply kong/api/endpoints.lua:328]: I hope I did it the right way and wanted :smile:

File: spec/02-integration/04-admin_api/10-services_routes_spec.lua [hunk 4]
```diff
@@ -191,6 +191,14 @@ for _, strategy in helpers.each_strategy() do
               pages[i] = json
             end
           end)
+          it("propagate in next a page size", function()
```
Comment (chronolaw) [inline spec/02-integration/04-admin_api/10-services_routes_spec.lua:194]: style: add a blank line before `it(...)`

Comment (jacksjm) [reply spec/02-integration/04-admin_api/10-services_routes_spec.lua:195]: Fixed :smile_cat:

**Review by fffonion — APPROVED** [review]

**Comment by fffonion** [comment]
@jacksjm  Thanks for the contribution, this is now merged 🎉 !

**Comment by jacksjm** [comment]
> @jacksjm  Thanks for the contribution, this is now merged 🎉 !

I'm very glad to contribute!!!

**Pull request merged**